api/src/trace/SugaredTracer.ts

```diff
@@ -0,0 +1,232 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import type {Context, Span, SpanOptions, Tracer} from '../';
+import * as api from '../';
+import {SpanStatusCode} from '../';
+
+const defaultOnException = (e: Error, span: Span) => {
+  span.recordException(e);
+  span.setStatus({
+    code: SpanStatusCode.ERROR
+  });
+};
+
+/**
+ * return a new SugaredTracer created from the supplied one
+ * @param tracer
+ */
+export function wrapTracer(tracer: Tracer): SugaredTracer {
+  return new SugaredTracer(tracer);
+}
+
+class SugaredTracer implements Tracer {
+  _tracer: Tracer;
+
+  constructor(tracer: Tracer) {
+    this._tracer = tracer;
+    this.startSpan = tracer.startSpan.bind(this._tracer);
+    this.startActiveSpan = tracer.startActiveSpan.bind(this._tracer);
+  }
+
+  startActiveSpan: Tracer['startActiveSpan'];
+  startSpan: Tracer['startSpan'];
+
+  /**
+   * Starts a new {@link Span} and calls the given function passing it the
+   * created span as first argument.
+   * Additionally, the new span gets set in context and this context is activated
+   * for the duration of the function call.
+   * The span will be closed after the function has executed.
+   * If an exception occurs, it is recorded, the status is set to ERROR and the exception is rethrown.
+   *
+   * @param name The name of the span
+   * @param [options] SpanOptions used for span creation
+   * @param [context] Context to use to extract parent
+   * @param [onException] function to overwrite default exception behaviour
+   * @param fn function called in the context of the span and receives the newly created span as an argument
+   * @returns return value of fn
+   * @example
+   *     const something = tracer.withActiveSpan('op', span => {
+   *      // do some work
+   *     });
+   * @example
+   *     const something = await tracer.withActiveSpan('op', span => {
+   *      // do some async work
+   *     });
+   */
+  withActiveSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    fn: F,
+  ): ReturnType<F>
+  withActiveSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SpanOptions,
+    fn: F,
+  ): ReturnType<F>
+  withActiveSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SpanOptions | F,
+    context: Context | F,
+    fn: F,
+  ): ReturnType<F>
+  withActiveSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SpanOptions,
+    context: Context,
+    onException: (e: Error, span: Span) => void,
+    fn: F,
+  ): ReturnType<F>
+  withActiveSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    arg2: SpanOptions | F,
+    arg3?: Context | F,
+    arg4?: ((e: Error, span: Span) => void) | F,
```

Comment:
I'd be cautious about adding new dedicated positional parameters to the span creation parameter list, as it may cause conflicts if we are going to extend the `Tracer.startSpan` parameter list (although I believe it would be unlikely to do so). 

What do you think about extending `SpanOptions` and moving the `onException` into that option bag?

Reply:
I think we should not extend `SpanOptions` itself but create an interface which derives from it. e.g. `SugaredSpanOptions`.

Reply:
Yeah, by saying "extending" I mean creating a new interface that extends the `SpanOptions`.